### PR TITLE
performance improvements, 4.8.1 release candidate

### PR DESCRIPTION
**Pull request by landreev**
## New Contributors

Welcome! New contributors should at least glance at [CONTRIBUTING.md](/CONTRIBUTING.md), especially the section on pull requests where we encourage you to reach out to other developers before you start coding. Also, please note that we measure code coverage and prefer you write unit tests. Pull requests can still be reviewed without tests or completion of the checklist outlined below. Thanks!

## Related Issues

- connects to #4173 abysmal performance of the dataset page
- connects to #4183 [hmdc.harvard.edu #253782] Unable to verify email address

## Pull Request Checklist

- [ ] Unit [tests][] completed
- [ ] Integration [tests][]: None
- [ ] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [ ] [Documentation][docs] completed
- [ ] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: https://github.com/IQSS/dataverse/tree/develop/scripts/database/upgrades
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/4.6.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/DatasetPage.java [hunk 1]
```diff
@@ -502,31 +487,46 @@ public boolean isSwiftStorage() {
...
+        }
+        
         if (isSwiftStorage() && (settingsService.getValueForKey(SettingsServiceBean.Key.ComputeBaseUrl) != null)) {
-            return true;
+            showComputeButtonForDataset = true;
+        } else {
+            showComputeButtonForDataset = false;
         }
-        return false;
+        return showComputeButtonForDataset;
```
Comment (oscardssmith) [inline src/main/java/edu/harvard/iq/dataverse/DatasetPage.java]: Why did you change this? At the very least, you should, change it to `showComputeButtonForDataset =(isSwiftStorage() && (settingsService.getValueForKey(SettingsServiceBean.Key.ComputeBaseUrl) != null))
return showComputeButtonForDataset`

File: src/main/java/edu/harvard/iq/dataverse/DatasetPage.java [hunk 2]
```diff
@@ -502,31 +487,46 @@ public boolean isSwiftStorage() {
...
     public boolean showComputeButton(FileMetadata metadata) {
+        if (showComputeButtonForFile != null) {
+            return showComputeButtonForFile;
+        }
+        
         if (isSwiftStorage(metadata) && (settingsService.getValueForKey(SettingsServiceBean.Key.ComputeBaseUrl) != null)) {
-            return true;
+            showComputeButtonForFile = true;
+        } else {
+            showComputeButtonForFile = false;
```
Comment (oscardssmith) [inline src/main/java/edu/harvard/iq/dataverse/DatasetPage.java]: same here

File: src/main/java/edu/harvard/iq/dataverse/dataaccess/ImageThumbConverter.java [hunk 3]
```diff
@@ -434,59 +428,108 @@ private static boolean isThumbnailCached(StorageIO<DataFile> storageIO, int size
...
+
+        if (storageIO == null) {
+            return null;
+        }
+
+        // skip the "isAvailable()" check - and just try to open the cached object. 
+        // if we can't open it, then we'll try to generate it. In other words, we are doing it in 
+        // the reverse order - and his way we can save one extra lookup, for a thumbnail 
+        // that's already cached - and on some storage media (specifically, S3)
+        // lookups are actually more expensive than reads.
```
Comment (oscardssmith) [inline src/main/java/edu/harvard/iq/dataverse/dataaccess/ImageThumbConverter.java]: what's the difference between a lookup and a read?

File: src/main/java/edu/harvard/iq/dataverse/dataaccess/S3AccessIO.java [hunk 4]
```diff
@@ -521,29 +522,39 @@ public OutputStream getOutputStream() throws UnsupportedDataAccessOperationExcep
...
+            if (key == null) {
+                String baseKey = this.getDataFile().getOwner().getAuthority() + "/" + this.getDataFile().getOwner().getIdentifier();
+                String storageIdentifier = dvObject.getStorageIdentifier();
+
+                if (storageIdentifier == null || "".equals(storageIdentifier)) {
+                    throw new FileNotFoundException("Data Access: No local storage identifier defined for this datafile.");
+                }
+
+                if (storageIdentifier.startsWith("s3://")) {
+                    bucketName = storageIdentifier.substring(storageIdentifier.indexOf(":") + 3, storageIdentifier.lastIndexOf(":"));
```
Comment (oscardssmith) [inline src/main/java/edu/harvard/iq/dataverse/dataaccess/S3AccessIO.java]: If you are going to have a hard +3, why not just assume that the index will be 6?

File: src/main/java/edu/harvard/iq/dataverse/DatasetPage.java (hunk 1 above)
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/DatasetPage.java]: But that's exactly what it's doing now, no? 
The reason I changed it in the first place was to only run the test once, then cache the result and return it the next time the page asks. (The problem with this method was that it was implemented in a super expensive way; AND the page calls it 24 times (!) as it's loading).

File: src/main/java/edu/harvard/iq/dataverse/DatasetPage.java (hunk 2 above)
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/DatasetPage.java]: OK, yeah - I screwed up with this one. The idea was to do the same thing as with the method above; to check it once, then cache the result. I just realized that this one is called on *different* files. So I can't just have one boolean to cache the result (doh). In order to cache these results I would need a Map of FileId-to-booleans. But the thing is, this probably doesn't need to be cached in the first place. Since this test is actually cheap.

File: src/main/java/edu/harvard/iq/dataverse/DatasetPage.java (hunk 1 above)
Comment (oscardssmith) [reply src/main/java/edu/harvard/iq/dataverse/DatasetPage.java]: the only difference is a useless if else statement. Rather than saying `if x{y=true}else{y=false}`, just having 'y=x` is both simpler and faster.

File: src/main/java/edu/harvard/iq/dataverse/dataaccess/ImageThumbConverter.java (hunk 3 above)
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/dataaccess/ImageThumbConverter.java]: "check if the object exists", "retrieve the object's metadata, such as size, etc." - these are lookups. As opposed to reading the actual bytes associated with the object. 
In the case of S3 this is a lookup:
size = s3.getObjectMetadata(bucketName,destinationKey).getContentLength();

and this is a read: 

S3Object s3object = s3.getObject(new GetObjectRequest(bucketName, destinationKey));
InputStream in = s3object.getObjectContent();
in.read(...); etc. 

With the thumbnails on the dataset page - which are bunch of tiny files that we need to read to load the page - all the extra lookups - "is the main file stored?", "is the thumbnail already cached?" (we were doing this lookup more than once too, apparently), "what is the size of the cached thumbnail?" were adding up to a surprising amount of extra time. Just going directly for the bytes improved it significantly.

File: src/main/java/edu/harvard/iq/dataverse/dataaccess/S3AccessIO.java (hunk 4 above)
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/dataaccess/S3AccessIO.java]: I copy-and-pasted this fragment from somewhere else in the class. And I'm assuming it was originally copy-and-pasted from another place, where this piece of code was run on storage identifiers with different prefixes (not just "s3:", but also "file: and "swift:", etc.). So the idea was that the length of "://" is always 3; but the length of "{prefix}://" can be anything. 
But yeah, the real crime with the fragment above is that it's now duplicated in 2 places in the class; I was just trying to quickly make it possible to read aux files without doing the full .open() on the object. I'll clean it up.

File: src/main/java/edu/harvard/iq/dataverse/DatasetPage.java (hunk 1 above)
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/DatasetPage.java]: I use verbose, multi-line if / else because it's easier to read. I know it's subjective.
